### ISSUE TITLE
Touchscreen events don't work

### ISSUE THREAD

**Issue by intirix** [anbox/anbox#25]
Make sure you are running the latest version of Anbox before reporting an issue.

**Anbox release (`anbox version`) and system info (`anbox system-info`):**
anbox 0.1.0

version: 0.1.0
os:
  name: Ubuntu Core
  version: 16
  snap-based: true
kernel:
  version: Linux version 4.4.0-72-generic (buildd@lcy01-17) (gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.4) ) #93-Ubuntu SMP Fri Mar 31 14:07:41 UTC 2017
  binder: true
  ashmem: true
graphics:
  egl:
    vendor: Mesa Project
    version: 1.4 (DRI2)
    extensions:
      - EGL_CHROMIUM_sync_control
      - EGL_EXT_create_context_robustness
      - EGL_EXT_image_dma_buf_import
      - EGL_KHR_create_context
      - EGL_KHR_fence_sync
      - EGL_KHR_get_all_proc_addresses
      - EGL_KHR_gl_renderbuffer_image
      - EGL_KHR_gl_texture_2D_image
      - EGL_KHR_gl_texture_cubemap_image
      - EGL_KHR_image
      - EGL_KHR_image_base
      - EGL_KHR_image_pixmap
      - EGL_KHR_reusable_sync
      - EGL_KHR_surfaceless_context
      - EGL_KHR_wait_sync
      - EGL_MESA_configless_context
      - EGL_MESA_drm_image
      - EGL_MESA_image_dma_buf_export
      - EGL_NOK_swap_region
      - EGL_NOK_texture_from_pixmap
      - EGL_NV_post_sub_buffer
      - EGL_WL_bind_wayland_display
  gles2:
    vendor: Intel Open Source Technology Center
    vendor: OpenGL ES-CM 1.1 Mesa 12.0.6
    extensions:
      - GL_EXT_blend_minmax
      - GL_EXT_multi_draw_arrays
      - GL_EXT_texture_filter_anisotropic
      - GL_EXT_texture_lod_bias
      - GL_OES_byte_coordinates
      - GL_OES_fixed_point
      - GL_OES_stencil_wrap
      - GL_OES_compressed_paletted_texture
      - GL_OES_query_matrix
      - GL_OES_read_format
      - GL_OES_single_precision
      - GL_EXT_texture_compression_dxt1
      - GL_OES_draw_texture
      - GL_OES_point_size_array
      - GL_OES_point_sprite
      - GL_EXT_texture_format_BGRA8888
      - GL_OES_compressed_ETC1_RGB8_texture
      - GL_OES_depth24
      - GL_OES_element_index_uint
      - GL_OES_fbo_render_mipmap
      - GL_OES_framebuffer_object
      - GL_OES_mapbuffer
      - GL_OES_rgb8_rgba8
      - GL_OES_stencil8
      - GL_OES_texture_env_crossbar
      - GL_OES_texture_mirrored_repeat
      - GL_OES_texture_npot
      - GL_OES_EGL_image
      - GL_OES_packed_depth_stencil
      - GL_OES_texture_cube_map
      - GL_APPLE_texture_max_level
      - GL_EXT_discard_framebuffer
      - GL_EXT_read_format_bgra
      - GL_OES_blend_equation_separate
      - GL_OES_blend_func_separate
      - GL_OES_blend_subtract
      - GL_OES_EGL_image_external
      - GL_OES_EGL_sync
      - GL_OES_vertex_array_object
      - GL_ANGLE_texture_compression_dxt3
      - GL_ANGLE_texture_compression_dxt5
      - GL_EXT_map_buffer_range
      - GL_KHR_debug
      - GL_OES_surfaceless_context
      - GL_EXT_polygon_offset_clamp

**Description of the problem:**

Touch events don't seem to work.  The interface changes colors sometimes, but the touch events don't seem to do anything

**Expected:**

Users should be able to use a touch screen to interact with an Android app

**Additional info:**

Great work!

Dell Inspiron 3157

**Comment by morphis** [comment]
Yeah this is a known limitation. If you look at https://github.com/anbox/anbox/blob/master/src/anbox/ubuntu/platform_policy.cpp#L74 you will see that we currently only handle pointer and keyboard devices. This needs to be extended to also support touchscreen type devices. For people looking at the code also notice the changes we did on the Android side for this:

 * https://github.com/anbox/platform_frameworks_native/commit/bf7ef8d3307d1822720cfc0305e48c0b9212c912
 * https://github.com/anbox/platform_frameworks_native/commit/d1df212bd1b776d9cb278a7f20cbfac93da2d54f
 * https://github.com/anbox/platform_frameworks_native/commit/1ec301bf2e59fdf94a21a61be85db26bb4f1980c

Someone who wants to fix the code has to add a touchscreen type device in Anbox itself and also needs to make sure its properly detected on the Android side in parallel to the pointer and keyboard devices.

**Comment by fhteagle** [comment]
I would be happy to test this enhancement if/when anyone works on the code. Code is not my forte, but happy to debug. Ubuntu 18.04 Gnome and Enlightenment....

**Comment by floe** [comment]
I do have a properly working Linux multitouch device, and would be very interested in implementing this. @morphis could you give me some up-to-date pointers regarding where to start, i.e. what is the input event flow?

**Comment by floe** [comment]
Addendum: Anbox edge (157) is working properly on my device, just not with touchscreen support. And I wonder if #923 is also related?

**Comment by sans-ltd** [comment]
Since lotxs of Android apps benefit of touchscreens and touch gestures (or are even unusable without them e.g. navigation apps) touch and gesture support would be a big benefit. 
Especially on convertibles like the Lenovo Yogas.

**Comment by uudruid74** [comment]
+1 as well.

I'm surprised that mouse support was implemented first as opposed to emulating a touch screen on the Android side with a mouse.  The touch screen should have been high priority.

**Comment by marcemira** [comment]
Yes, this should be a must have in order to get most applications working closer to real daily usage. Is there any software that can emulate touch events from mouse events working on Linux currently? that'll be very helpful in the mean time. Thanks

**Comment by ZZZXXX0110** [comment]
Any update on this issue lately?

**Comment by marcemira** [comment]
Experimenting a bit, I was able to send and execute `input` terminal commands through adb to simulate touch events on a full screen, so it should be very possible to have a little script that(with a bit more effort) could get the active window dimensions and send the simulated touch events with this method, probably easier start with full screen and single screen first.

**Comment by uudruid74** [comment]
This makes no sense.  Android.x86 already supports touch screens.  Anbox just needs to allow access to the device.

**Comment by marcemira** [comment]
@uudruid74 but then you'll need a touch screen device to send those events. If you just have a mouse and you want to use apps that requires touch events only, that needs to be emulated, either internally by android or with a helper tool specifically.

**Comment by uudruid74** [comment]
Android already allows the use of a mouse and we have no issues with mouse.  Emulating a touch screen shouldn't be necessary at all, and if you need to, that's an entirely different issue and should have a different bug filed.

This issue is that people with a touchscreen can't use it (with an OS designed to use a touchscreen as it's primary input device).  Let's not bastardize an implementation trying to solve other issues not at hand

**Comment by ZZZXXX0110** [comment]
Precisely, this issue is about being able to use Anbox on a device that
already has a touch screen, aka a touchscreen device. Currently on a
touchscreen device, touchscreen input is basically ignored by An box.

**Comment by sans-ltd** [comment]
Multitouch is actually essential for map/navigation apps which are not available natively on Linux at all. So anbox is a viewable option to get Navigation under Linux working. Currently essential gestures like pinch-zoom or flick/swipe Scrolling are not possible at all. 
All recent Notebooks have multitouch capable touchpads and a lot of Notebooks/convertibles today have Touchscreens now.

**Comment by uudruid74** [comment]
Don't have a clue what you mean.  My punch gestures work fine.  Again, if you need to emulate a touchscreen, thats a different issue.  Make a feature request.   All the gestures you mention work just fine natively

**Comment by ZZZXXX0110** [comment]
Do you have a touchscreen? What DE are you using? On my Surface Pro 6 all
touch screen input is ignored by Anbox.

**Comment by uudruid74** [comment]
Yes, ignored by Anbox, NOT by Linux
  
  
  

  
  
  
  
  
>

**Comment by ZZZXXX0110** [comment]
Any update on this? It seems like we already have a working solution in the form of a pull request?

**Comment by ZZZXXX0110** [comment]
Touch event is still not working on the latest build from edge channel (version 4-d521e28) :(

**Comment by salvprest** [comment]
this should be solved now

**Comment by fhteagle** [comment]
I can confirm single touch events are recognized in anbox (snap rev 176) in Kubuntu 19.04. Multitouch does not appear to be working. There is a slight chance that is an issue with my KDE Plasma /  X11 setup. Can anyone else confirm that multitouch events are not passed on to / utilized by anbox?

**Comment by tuxflo** [comment]
There is an open pull request that adds multitouch functionality to anbox: https://github.com/anbox/anbox/pull/1129
It works here on my Lenovo Yoga X260 with a Wacom Touch screen running KDE neon (Ubuntu 18.04 with latest KDE software, X11).
However it is still kind of hacky to emulate touch events with SDL instead of passing the devices directly to the android layer. SDL also has limitations in terms of Stylus handling. My stylus is not detected by SDL so using it with anbox isn't possible :(

**Comment by theWebalyst** [comment]
Touch screen not working with:
- AnBox 4
- Ubuntu 19.04
- Gnome desktop 3.32.1
- DrawExpress Diagram Lite 2.0.6
- Dell Laptop Screen: UltraSharp™ FHD IPS Touch

I installed AnBox because I wanted to use the awesome gesture input on DrawExpress.

The app works fairly well using the mouse for input but there are some issues there too, see below.

Using the touch screen there is some response from the app, but no functionality is available:
- Drawing gestures have no effect (e.g. if I draw a circle with my finger on the screen, it should show the path until I finish the shape, and then create a circle object on the diagram).
- Tapping my finger on the menu icons causes them to lighten in colour, but none of them show their menus.
- Similarly, the AnBox and DrawExpress windows don't respond to the touch screen. Tapping on icons and trying to click on or drag the window bar have no effect (e.g. I can drag other windows to move them around the screen by their title bar).

So no app functionality is available through my touch screen in AnBox or Android apps.

There are also issues with mouse input in DrawExpress:
- I can use the mouse input gestures to create a shape, and some but not all of the menu icons respond reliably to mouse clicks. The colour palette, undo/redo seem ok, but the main menu icon doesn't open unless you click near the edge of the icon!
- having clicked on the lock (padlock) icon, the current diagram is effectively broken because instead of unlocking the diagram, clicking on the padlock closes the DrawExpress app (and often also closes AnBox too). Ctrl-Enter closes the now uneditable diagram, so I can at least use the app again, but the diagram is no longer editable.

**Comment by angelovAlex** [comment]
Have the same issues as @theWebalyst. Touch screen doesn’t click, and looks like anbox handles touch events as ‘mouse move’, as it highlights buttons when pressing on buttons, but no actions happen.
